### PR TITLE
[enhancement] Add ability to integrate sanitizers in oneDAL Make builds

### PR DESCRIPTION
**Pull request by icfaust**
## Description

This adds the ```REQSAN``` flag to the make build which can trigger dynamic linking of an available sanitizer into oneDAL.  This will allow for enhanced memory leak and race condition checking among other possibilities. These are determined from the compiler, and is expected to be the responsibilty of the developer to chose a sanitizer which is available. The most standard practice should be to use ```REQDBG=yes REQSAN=address```

The only special case is value ```static``` which will integrate ASan and link its static version, which requires a special linker flag ```-static-libasan```

No performance benchmarks necessary.

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.
- [x] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

File: makefile [hunk 1]
```diff
@@ -139,6 +139,8 @@ y           := $(notdir $(filter $(_OS)/%,lnx/so win/dll mac/dylib))
 -DEBC       := $(if $(REQDBG),$(if $(filter symbols,$(REQDBG)),$(-DEBC.$(COMPILER)),$(-DEBC.$(COMPILER)) -DDEBUG_ASSERT -DONEDAL_ENABLE_ASSERT)) -DTBB_SUPPRESS_DEPRECATED_MESSAGES -D__TBB_LEGACY_MODE
 -DEBC_DPCPP := $(if $(REQDBG),$(if $(filter symbols,$(REQDBG)),$(-DEBC.dpcpp),$(-DEBC.dpcpp) -DDEBUG_ASSERT -DONEDAL_ENABLE_ASSERT))
 -DEBL       := $(if $(REQDBG),$(if $(OS_is_win),-debug,))
+-sanitize   := $(if $(REQASAN),$(if $(COMPILER_is_vc),/fsanitize=address,-fsanitize=address))
+-lsanitize  := $(-sanitize) $(if $(REQASAN),$(if $(filter static,$(REQASAN)),-static-libasan))
```
Comment (david-cortes-intel) [inline makefile:143]: Shouldn't this be `-static-libsan` for GCC?

**Comment by david-cortes-intel** [comment]
@icfaust Great addition. Since the compiler we use (ICX) is based off clang, how about adding also MSAN and TSAN when supported? They cannot be all added together in the same build, but could also be offered as makefile options.

**Comment by david-cortes-intel** [comment]
@icfaust I believe this would also require a similar option for the `setup.py` of sklearnex. I think adding the path with the `libasan.so` file to the rpath should suffice but not 100% sure.

File: makefile (hunk 1 above)
Comment (icfaust) [reply makefile:143]: Good question, some of the documentation online said that there was a difference between GCC and Clang on ```-static-libasan``` vs ```-static-libsan```. Local testing and GCC documentation have shown that ```-static-libasan``` is supported: https://gcc.gnu.org/onlinedocs/gcc/Link-Options.html

File: makefile [hunk 2]
```diff
@@ -139,6 +139,8 @@ y           := $(notdir $(filter $(_OS)/%,lnx/so win/dll mac/dylib))
 -DEBC       := $(if $(REQDBG),$(if $(filter symbols,$(REQDBG)),$(-DEBC.$(COMPILER)),$(-DEBC.$(COMPILER)) -DDEBUG_ASSERT -DONEDAL_ENABLE_ASSERT)) -DTBB_SUPPRESS_DEPRECATED_MESSAGES -D__TBB_LEGACY_MODE
 -DEBC_DPCPP := $(if $(REQDBG),$(if $(filter symbols,$(REQDBG)),$(-DEBC.dpcpp),$(-DEBC.dpcpp) -DDEBUG_ASSERT -DONEDAL_ENABLE_ASSERT))
 -DEBL       := $(if $(REQDBG),$(if $(OS_is_win),-debug,))
+-sanitize   := $(if $(REQASAN),$(if $(COMPILER_is_vc),/fsanitize=address,-fsanitize=address))
```
Comment (david-cortes-intel) [inline makefile:142]: Shouldn't `REQSAN` also trigger it to add `-ggdb`? Otherwise the logs won't be very helpful.

Comment (icfaust) [reply makefile]: Good question, I saw some documentation with windows where /MDd etc wasn't playing well with ASan, so I wanted to leave it up to the user based on their use case. This also simplified how much testing I had to do/cleanly separate enabling ASan from things which impact debugging (i.e. ```REQDBG```).

**Comment by icfaust** [comment]
> @icfaust Great addition. Since the compiler we use (ICX) is based off clang, how about adding also MSAN and TSAN when supported? They cannot be all added together in the same build, but could also be offered as makefile options.

Added special options ```memory``` and ```thread```.

File: makefile (hunk 2 above)
Comment (icfaust) [reply makefile]: To get ASan with a build with debug symbols (no asserts) and static ASan:
```
REQDBG=symbols REQSAN=static
```

Comment (david-cortes-intel) [reply makefile]: Then would be ideal to put that in the example that it gives, or to mention that adding REQSAN will not automatically add debug symbols.

Comment (icfaust) [reply makefile]: Makes absolute sense, added to the main example and left a note that it is recommend to use ```REQDBG```.

**Comment by david-cortes-intel** [comment]
@icfaust After building with `REQSAN=yes` and  `REQDBG=yes`, I get this error when I try to build the C++ examples:
```
/export/users/dcortes/repos/oneDAL/__release_lnx/daal/latest/lib/intel64/libonedal_core.so.3: undefined reference to `__asan_set_shadow_f8'
```

I _think_ it should be possible to work around this by adding the right paths to `$LIBRARY_PATH` and `$LD_LIBRARY_PATH`, but not sure how to do this when installing with conda tooling or our internal tooling.

**Comment by david-cortes-intel** [comment]
OTOH, can at least confirm that it works with sklearnex. It requires forcefully LD_PRELOAD'ing the `libasan.so` object, but it shows logs as expected.

Example from KMeans:
```
Indirect leak of 440 byte(s) in 1 object(s) allocated from:
    #0 0x7fe3181b5667 in __interceptor_malloc (/lib64/libasan.so.6+0xb0667)
    #1 0x7fe2e62a98a6 in mm_internal_malloc (/localdisk2/mkl/dcortes/repos/oneDAL/__release_lnx/daal/latest/lib/intel64/libonedal_core.so.3+0x6d158a6)
    #2 0x7fe2dcb29590 in oneapi::dal::kmeans::v1::train_result<oneapi::dal::kmeans::task::v1::clustering> oneapi::dal::kmeans::backend::call_daal_kernel<double, oneapi::dal::kmeans::task::v1::clustering, oneapi::dal::kmeans::method::v1::lloyd_dense, oneapi::dal::v1::table>(oneapi::dal::backend::context_cpu const&, oneapi::dal::kmeans::detail::v1::descriptor_base<oneapi::dal::kmeans::task::v1::clustering> const&, oneapi::dal::v1::table const&, oneapi::dal::v1::table const&) cpp/oneapi/dal/algo/kmeans/backend/cpu/train_kernel_lloyd.cpp:103
    #3 0x7fe2dcb29590 in oneapi::dal::kmeans::v1::train_result<oneapi::dal::kmeans::task::v1::clustering> oneapi::dal::kmeans::backend::train<double, oneapi::dal::kmeans::task::v1::clustering, oneapi::dal::kmeans::method::v1::lloyd_dense>(oneapi::dal::backend::context_cpu const&, oneapi::dal::kmeans::detail::v1::descriptor_base<oneapi::dal::kmeans::task::v1::clustering> const&, oneapi::dal::kmeans::v1::train_input<oneapi::dal::kmeans::task::v1::clustering> const&) cpp/oneapi/dal/algo/kmeans/backend/cpu/train_kernel_lloyd.cpp:153
    #4 0x7fe2dcb29590 in oneapi::dal::kmeans::backend::train_kernel_cpu<double, oneapi::dal::kmeans::method::v1::lloyd_dense, oneapi::dal::kmeans::task::v1::clustering>::operator()(oneapi::dal::backend::context_cpu const&, oneapi::dal::kmeans::detail::v1::descriptor_base<oneapi::dal::kmeans::task::v1::clustering> const&, oneapi::dal::kmeans::v1::train_input<oneapi::dal::kmeans::task::v1::clustering> const&) const cpp/oneapi/dal/algo/kmeans/backend/cpu/train_kernel_lloyd.cpp:161
    #5 0x7fe2dc82cf9c in auto oneapi::dal::backend::kernel_dispatcher<oneapi::dal::backend::kernel_spec<oneapi::dal::backend::single_node_cpu_kernel, oneapi::dal::kmeans::backend::train_kernel_cpu<double, oneapi::dal::kmeans::method::v1::lloyd_dense, oneapi::dal::kmeans::task::v1::clustering> > >::operator()<oneapi::dal::kmeans::detail::v1::descriptor_base<oneapi::dal::kmeans::task::v1::clustering> const&, oneapi::dal::kmeans::v1::train_input<oneapi::dal::kmeans::task::v1::clustering> const&>(oneapi::dal::detail::v1::host_policy const&, oneapi::dal::kmeans::detail::v1::descriptor_base<oneapi::dal::kmeans::task::v1::clustering> const&, oneapi::dal::kmeans::v1::train_input<oneapi::dal::kmeans::task::v1::clustering> const&) const cpp/oneapi/dal/backend/dispatcher.hpp:185
    #6 0x7fe2dc82cf9c in oneapi::dal::kmeans::detail::v1::train_ops_dispatcher<oneapi::dal::detail::v1::host_policy, double, oneapi::dal::kmeans::method::v1::lloyd_dense, oneapi::dal::kmeans::task::v1::clustering>::operator()(oneapi::dal::detail::v1::host_policy const&, oneapi::dal::kmeans::detail::v1::descriptor_base<oneapi::dal::kmeans::task::v1::clustering> const&, oneapi::dal::kmeans::v1::train_input<oneapi::dal::kmeans::task::v1::clustering> const&) const cpp/oneapi/dal/algo/kmeans/detail/train_ops.cpp:31
    #7 0x7fe2dd7b2823  (/localdisk2/mkl/dcortes/repos/scikit-learn-intelex/onedal/_onedal_py_host.cpython-312-x86_64-linux-gnu.so+0x9d823)
```

File: makefile [hunk 3]
```diff
@@ -139,6 +139,11 @@ y           := $(notdir $(filter $(_OS)/%,lnx/so win/dll mac/dylib))
 -DEBC       := $(if $(REQDBG),$(if $(filter symbols,$(REQDBG)),$(-DEBC.$(COMPILER)),$(-DEBC.$(COMPILER)) -DDEBUG_ASSERT -DONEDAL_ENABLE_ASSERT)) -DTBB_SUPPRESS_DEPRECATED_MESSAGES -D__TBB_LEGACY_MODE
 -DEBC_DPCPP := $(if $(REQDBG),$(if $(filter symbols,$(REQDBG)),$(-DEBC.dpcpp),$(-DEBC.dpcpp) -DDEBUG_ASSERT -DONEDAL_ENABLE_ASSERT))
 -DEBL       := $(if $(REQDBG),$(if $(OS_is_win),-debug,))
+# NOTE: only some compilers support other sanitizers, failure is expected by design in order to not
+# quietly hide the lack of support (e.g. gnu will fail with REQSAN=memory). The default is AddressSanitizer
+# as it is available by all compilers which build oneDAL.
+-sanitize   := $(if $(REQSAN),$(if $(COMPILER_is_vc),/fsanitize=,-fsanitize=)$(if $(filter memory,$(REQSAN)),memory,$(if $(filter thread,$(REQSAN)),thread,address)))
+-lsanitize  := $(-sanitize) $(if $(REQSAN),$(if $(filter static,$(REQSAN)),-static-libasan))
```
[the remark this answers is not in the document]
Comment (icfaust) [reply makefile]: Changed design to enable all available sanitizers, now standard ASan must be specified with ```REQSAN=address``` with the only special case is ```static``` which will statically link ASan. Otherwise ```REQSAN``` must be an available sanitizer for the specific compiler, meaning that the whole range of Clang sanitizers will then be available (albeit without static linking).

File: INSTALL.md [hunk 4]
```diff
@@ -170,6 +170,18 @@ On **Linux\*** it is possible to build debug version of oneDAL or the version th
...
+
+- To build oneDAL with [AddressSanitizer](https://github.com/google/sanitizers/wiki/addresssanitizer) integration, run:
+
+    _Note: Windows support of ASan in oneDAL is experimental, AddressSanitizer is the default, and Debug symbols are recommended_
+
+            make -f makefile daal oneapi_c PLAT=lnx32e REQSAN=yes REQDBG=yes
+
+- To build oneDAL with [MemorySanitizer](https://github.com/google/sanitizers/wiki/memorysanitizer), run:
+
+    _Note: Only available on Clang and Clang-derived compilers_
```
[the remark this answers is not in the document]
Comment (icfaust) [reply INSTALL.md]: Change made (which specified DPC++ explicitly)

File: makefile (hunk 3 above)
[the remark this answers is not in the document]
Comment (david-cortes-intel) [reply makefile]: Thanks. But if you're aiming for the _whole range_ of sanitizers, there's also ubsan (but I have a feeling that it will report lots of issues that aren't actual issues due to the way the oneDAL codebase uses intel-specific tricks) and leaksan (a subset of asan). Plus `ASAN_OPTIONS` and similar ones for the other sanitizers.

File: INSTALL.md [hunk 5]
```diff
@@ -170,6 +170,20 @@ On **Linux\*** it is possible to build debug version of oneDAL or the version th
...
+
+- To integrate [AddressSanitizer](https://github.com/google/sanitizers/wiki/addresssanitizer) in a debug oneDAL build (recommended), run:
+
+    _Note: Windows support of REQSAN in oneDAL is experimental, static AddressSanitizer can be set with value: static_
+
+            make -f makefile daal oneapi_c PLAT=lnx32e REQSAN=address REQDBG=yes
+
+- To integrate [MemorySanitizer](https://github.com/google/sanitizers/wiki/memorysanitizer) in a debug oneDAL build, run:
+
+    _Note: Clang and Clang-derived compilers support MSan and TSan (including the Intel DPC++ compiler)_
```
Comment (david-cortes-intel) [inline INSTALL.md:184]:
```suggestion
    _Note: Clang and Clang-derived compilers (including the Intel DPC++ compiler) support additional sanitizers such MSan, TSan, and UBSan_
```

File: makefile [hunk 6]
```diff
@@ -1129,6 +1139,12 @@ Flags:
       except value "symbols" which will only add debug symbols.
       special value: symbols
   REQPROFILE - flag that enables kernel profiling using <ittnotify.h>
+  REQSAN - flag that integrates a Google sanitizer (e.g. AddressSanitizer).
+      The sanitizer must be specified as the flag value, except for the
+      value of "static". This value will link with the static ASan library.
+      Use on Windows is unverified, with the available sanitizers dependent
+      on the compiler. It is recommended to use in tandem with REQDBG.
+      special values: static
```
Comment (david-cortes-intel) [inline makefile:1147]:
```suggestion
      special values: static, address, memory, thread, leak, undefined
```

**Review by david-cortes-intel — APPROVED** [review]
